src/dataflow/src/source/timestamp.rs

```diff
@@ -165,35 +280,77 @@ impl TimestampBindingBox {
         }
     }
 
-    fn add_binding(&mut self, partition: PartitionId, timestamp: Timestamp, offset: MzOffset) {
-        let partition = self
-            .partitions
-            .entry(partition)
-            .or_insert_with(PartitionTimestamps::new);
-        partition.add_binding(timestamp, offset);
+    fn add_partition(&mut self, partition: PartitionId) {
+        if self.partitions.contains_key(&partition) {
+            debug!("already inserted partition {:?}, ignoring", partition);
+            return;
+        }
+
+        self.partitions
+            .insert(partition, PartitionTimestamps::new());
+    }
+
+    fn add_binding(
+        &mut self,
+        partition: PartitionId,
+        timestamp: Timestamp,
+        offset: MzOffset,
+        proposed: bool,
```

Comment:
nit: The `proposed` parameter feels a bit iffy. We're special casing the BYO and RT cases here. There's basically two call sites of this, one with `true` and one with `false`.